Android: Add Paths to UI

Ebola16 opened the pull request:
Implements https://bugs.dolphin-emu.org/issues/10781
- Android: Add Set as Default ISO to UI
- Android: Add Paths to UI
- Very useful for loading game mods
- Removed some unnecessary import statements
- Renamed some things now that `FilePicker` is more versatile
- Added `ConfirmRunnable` to confirm running arbitrary runnables (so they're not accidentally clicked). It will also be useful for future PRs, especially when modifying multiple settings (i.e. toggling all log types).
- Fixed a bug where `launchOpenFileActivity()` could select multiple files. We don't want to be able to load multiple games at once!

- [X] Review requested: See the TODO added to `SettingsActivity`. Comment out the `finish()` to see the weirdness. My workaround works but it probably isn't the prettiest!

![Screenshot_20200105-170115_Dolphin Emulator](https://user-images.githubusercontent.com/17330088/71786806-04eb5c00-2fde-11ea-851e-61f0a2253d87.jpg)

![Screenshot_20200322-054716_Dolphin Emulator](https://user-images.githubusercontent.com/17330088/77246958-b6c01f00-6c02-11ea-9a79-fe215ac37161.jpg)

JosJuice commented:
I don't think we should have this without having the default ISO exposed in the settings, since there otherwise would be no way to clear it from the GUI.

JosJuice commented on Source/Android/app/src/main/res/values/arrays.xml:
```diff
@@ -326,12 +326,14 @@
         <item>GFX Settings</item>
         <item>GameCube Controller Settings</item>
         <item>Clear Game Settings</item>
+        <item>Set as default ISO</item>
```
The capitalization doesn't match the other options.

Ebola16 replied:
Think I should capitalize "As" too?

JosJuice replied:
No.

Ebola16 commented:
There's no play button on Android so is there a reason to clear the default ISO setting? Other than the play button, I think only mod launchers care about this setting and it can now be easily changed from Dolphin's UI. But we could wait on this PR until the `Paths` submenu is implemented.

leoetlino commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/GameSettingsDialog.java:
```diff
@@ -64,14 +68,24 @@ public Dialog onCreateDialog(Bundle savedInstanceState)
                   break;
                 case 4:
-                  clearGameSettings(gameId);
+                  // Set as default ISO for GC, Wii clear game settings for else
```
This seems very error prone to me... Is there no other way to tell what option was selected?  Not a big fan of those unnamed indices which are essentially magic numbers that can even change depending on the game file platform (!)

Ebola16 replied:
Yeah, I think I found a better method.

Ebola16 replied:
I moved my changes to earlier in the array to reduce ugliness. The cases aren't named because they change depending upon platform. I think the goal was to conditionally include Wiimote settings while keeping Clear Game Settings at the end of the array. I could separate the Game Properties `.setItems` by platform but then I'd need to duplicate the `switch`. There's only one case that differs between platforms so I'm thinking I should keep this as-is. Let me know if you disagree.

Ebola16 commented:
> I don't think we should have this without having the default ISO exposed in the settings, since there otherwise would be no way to clear it without editing Dolphin.ini.
I took care of @JosJuice's concerns. I expanded the scope of this PR to include the Paths submenu. See updated original description for noteworthy changes. Requesting a review!

Ebola16 commented:
> I don't think we should have this without having the default ISO exposed in the settings, since there otherwise would be no way to clear it without editing Dolphin.ini.

I took care of @JosJuice's concerns. I expanded the scope of this PR to include the Paths submenu. See updated original description for noteworthy changes. Requesting a review!

Ebola16 commented:
Slight comment clarification. Although the changes in `SettingsActivity` may be somewhat wonky, I don't think there's any drawbacks to this PR so ready for review!

Ebola16 commented:
I realize the few Android devs may be busy but ...anyone?

JosJuice commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/CustomFilePickerFragment.java:
```diff
@@ -21,7 +21,7 @@
 {
   private static final Set<String> extensions = new HashSet<>(Arrays.asList(
-          "gcm", "tgc", "iso", "ciso", "gcz", "wbfs", "wad", "dol", "elf", "dff"));
+          "gcm", "tgc", "iso", "ciso", "gcz", "wbfs", "wad", "dol", "elf", "dff", "raw"));
```
I'm assuming you did this because it's needed for the SD card path, but this set of extensions is also used for selecting a game to play. If we're going to filter by extension, which extensions are allowed should be dependent on what you're using the file picker for.

JosJuice commented on Source/Android/app/src/main/res/values/strings.xml:
```diff
@@ -162,6 +162,16 @@
 
+    <!-- Path Settings -->
+    <string name="paths_submenu">Paths</string>
+    <string name="search_subfolders_ISO">Search Subfolders for Game Files</string>
```
This string is not used.

Ebola16 replied:
Removed. Search subfolder's setting seems to do nothing so that will be saved for another PR.

Ebola16 commented:
I implemented clearing the default ISO as a runnable that asks the user to confirm the action (see third screenshot). This will be used in future PRs too (i.e. toggling all log types).

I agree that the behavior in `SettingsActivity` is weird but it's the best I can come up with at the moment. That covers two of the TODOs I added. With the most recent comment, I've added a third TODO in `SettingsFragment` for a minor graphical glitch. I'm open to suggestions for improvement, but I think this can be merged and I'll do follow-up PRs if I can address the TODOs. This PR is getting fairly large.

Ebola16 commented:
Also, fixed a bug where `launchOpenFileActivity()` could select multiple files. We don't want to be able to load multiple games at once!

Ebola16 commented:
Is this in an acceptable state to merge? I've documented the minor problems in TODOs but this PR functions as intended as far as I'm aware. I'd also like to use the features added here in future PRs.

I'm always open to suggestions to improve the code too, but I'd like this merged as the PR is becoming fairly large.

Ebola16 commented:
I think I resloved one of the TODOs. In `SettingsActivity`, calling ` mPresenter.onStop()` seems to be the best course of action and I added a check to prevent duplicate Toasts. I also removed some redundant code and unnecessary includes.

Ebola16 commented:
I think I resolved `SettingsFragment`'s TODO by adding a blank submenu and then immediately backing out of it to update dynamic settings descriptions.

This PR is more complicated than I expected!

Ebola16 commented:
...it seems nobody has any other comments on this PR so can it be merged?

JosJuice commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/settings/ui/SettingsActivityPresenter.java:
```diff
@@ -21,7 +21,7 @@
   private int mStackCount;
 
-  private boolean mShouldSave;
+  boolean mShouldSave;
```
Could we add a getter for this instead of changing the visibility?

JosJuice commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/settings/ui/SettingsAdapter.java:
```diff
@@ -274,6 +300,54 @@ public void onInputBindingClick(final InputBindingSetting item, final int positi
+  public static void onFilePickerConfirmation(String file)
+  {
+    NativeLibrary
+            .SetConfig(SettingsFile.FILE_NAME_DOLPHIN + ".ini", mItem.getSection(), mItem.getKey(),
```
Could this be made to not hardcode the filename, in case we want to use this for other settings in the future? (I don't understand the code structure here well enough to know how easy it would be, so if it would be too hard, don't worry about it.)

Ebola16 replied:
Yes, done.

JosJuice commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/settings/ui/SettingsAdapter.java:
```diff
@@ -57,6 +67,14 @@
   private TextView mTextSliderValue;
 
+  private static FilePicker mFilePicker;
+  private static SettingsItem mItem;
```
These should be prefixed with `s`, not `m`.

JosJuice commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/settings/ui/SettingsAdapter.java:
```diff
@@ -274,6 +300,54 @@ public void onInputBindingClick(final InputBindingSetting item, final int positi
+        return gameExtensions;
+      }
+    }
+    catch (NullPointerException ex)
```
Is there a situation where this gets triggered in practice? Assuming it can happen when calling `getRequestType()`, I feel it would be cleaner to just check if `mFilePicker` is `null`.

Ebola16 replied:
Yes, this is necessary. FilePickers called from the add directory button and "Open File" menu don't have extensions set in `SettingsFragmentPresenter`. These cases don't mix well with the static variables used in `getExtensions`. There are other cases in Dolphin where we use this method to handle possibly null int too (search for `NullPointerException ex`)

JosJuice commented:
I found a bug: If you first go to the paths settings and bring up the file picker for the SD card path (and press either Cancel or OK), and then go to add a game directory to the game list, the directory picker will only show .raw files instead of only game files.

The logic for which set of extensions to use seems quite fragile, with the static variable in SettingsAdapter. If it's possible to do it without static variables like that, that would be preferable.

Ebola16 commented:
The bug you described has been fixed by clearing the static variables in `SettingsActivity`. Unfortunately I think the static variables are necessary here to avoid multiple inheritance problems.
The latest version of this PR:
- Adds some static variables I accidentally removed in the previous version in `SettingsAdapter`. FilePicker should now properly interact with directory selection.
- Repurposed `clearDefaultISO` to `resetPaths` (see second image in original post). It's silly to just reset one path. If users are messing around with these settings we might as well easily restore them all.

Hopefully the latest version addresses everything. I think this is good to go and I received feedback so I'm removing my [Feedback Wanted] label.

JosJuice commented:
I have made a commit that makes CustomFilePickerFragment not rely on the static state in SettingsAdapter. If you could include it, or something similar to it, I think I would be happy with merging the PR.

I've checked that it works, but I haven't lint checked it, so maybe you'll need to edit the formatting of one or two lines.

https://github.com/JosJuice/dolphin/commit/779be103adb4f22a775bd45fff7ddfdb55d74827

Ebola16 commented:
@JosJuice I pushed one more commit because `onSettingsChanged` could still clobber changes made by `ConfirmRunnable`. The main functional changes are `ConfirmRunnable` now calls `finish` upon completion and yet another static variable has been added to `resetPaths`. This also allows me to remove the now unnecessary `reloadSubMenu` and associated changes.

Let me know if you want me to squash the last 3 commits.

Ebola16 commented:
@JosJuice can we get this merged? I have a feeling nobody else is going to review this PR.

Ebola16 commented:
@ds22x That's likely possible, although using variables in paths may confuse users. I don't know if I'll get around to implementing that but you can file a feature request at https://bugs.dolphin-emu.org/projects/emulator/issues so people can better see your suggestion.

ds22x commented:
And done.
https://bugs.dolphin-emu.org/issues/12046
